### PR TITLE
Time sync: TSC feature

### LINKED ISSUE #28271: [Feature] TSC for time sync
### Feature description

...
TSC impl is not yet done. We should do it.

### Platform

core

### Platform Version(s)

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by cecille**
Implements the time sync client, which allows devices to reach out to other matter nodes to query time.

Because this feature is challenging to test in the CI, this PR also includes a runtime options provider implementation that allows the time cluster to fake having no system time, forcing a read out over the network.

Tests: The added tests MUST be run using the provided script runner because they require a specific network and app setup to work correctly. The first two tests are used to prove the platform time bypass is functional (test_SimulateNoInternalTime and test_HaveInternalTime). The next "test" sets up a second node to act as the trusted time source. test_SetAndReadFromTrustedTimeSource tests that the node can read from the trusted time source immediately after setting it up.  test_ReadFromTrustedTimeSource tests that the values persist correct and the node can read from the trusted time source after reboot.

This PR also includes an expansion of the delegate API, which will allow platforms to supply their own time. Currently, this is done using the platform time abstractions to match the current style of the time sync server impl.

Setting gn arg time_sync_enable_tsc_feature to 0 will turn this feature off. Note that this does NOT control the zap feature map.

Fixes: #28271

### CODE REVIEW

**Review by jmartinez-silabs — APPROVED** [review]

**Review by aronrosenberg — APPROVED** [review]

File: src/app/clusters/time-synchronization-server/time-synchronization-server.cpp [hunk 1]
```diff
@@ -228,6 +267,186 @@ TimeSynchronizationServer & TimeSynchronizationServer::Instance()
...
+{
+    // Sent as a char-string to the delegate so they can read it easily
+    char defaultNTP[kMaxDefaultNTPSize];
+    MutableCharSpan span(defaultNTP);
+    if (GetDefaultNtp(span) != CHIP_NO_ERROR)
+    {
+        emitTimeFailureEvent(kRootEndpointId);
+        return;
+    }
+    if (span.size() > kMaxDefaultNTPSize)
```
Comment (bzbarsky-apple) [inline src/app/clusters/time-synchronization-server/time-synchronization-server.cpp:288]: How could this happen, short of the callee writing out of bounds (in which case we are already in trouble)?

File: src/app/clusters/time-synchronization-server/time-synchronization-server.cpp [hunk 2]
```diff
@@ -228,6 +267,186 @@ TimeSynchronizationServer & TimeSynchronizationServer::Instance()
...
+    {
+        emitTimeFailureEvent(kRootEndpointId);
+    }
+}
+
+#if TIME_SYNC_ENABLE_TSC_FEATURE
+void TimeSynchronizationServer::OnDeviceConnectedFn(Messaging::ExchangeManager & exchangeMgr, const SessionHandle & sessionHandle)
+{
+    // Connected to our trusted time source, let's read the time.
+    app::AttributePathParams readPaths[2];
```
Comment (bzbarsky-apple) [inline src/app/clusters/time-synchronization-server/time-synchronization-server.cpp:303]: Why the `app::` prefixes here?  This file has `using namespace chip::app`....

File: src/app/clusters/time-synchronization-server/time-synchronization-server.cpp [hunk 3]
```diff
@@ -228,6 +267,186 @@ TimeSynchronizationServer & TimeSynchronizationServer::Instance()
...
+                                            app::Clusters::TimeSynchronization::Attributes::UTCTime::Id);
+    readPaths[1] = app::AttributePathParams(kRootEndpointId, app::Clusters::TimeSynchronization::Id,
+                                            app::Clusters::TimeSynchronization::Attributes::Granularity::Id);
+
+    app::InteractionModelEngine * engine = app::InteractionModelEngine::GetInstance();
+    app::ReadPrepareParams readParams(sessionHandle);
+    readParams.mpAttributePathParamsList    = readPaths;
+    readParams.mAttributePathParamsListSize = 2;
+
+    auto attributeCache = Platform::MakeUnique<app::ClusterStateCache>(*this);
```
Comment (bzbarsky-apple) [inline src/app/clusters/time-synchronization-server/time-synchronization-server.cpp:314]: Why do you need a ClusterStateCache?  That's a bunch of code being pulled in with a bunch of codesize overhead that's totally not needed if we're just trying to read these two attributes....

The ClusterStateCache is written with high-resource controllers in mind, not constrained devices.

File: src/app/clusters/time-synchronization-server/time-synchronization-server.cpp [hunk 4]
```diff
@@ -228,6 +267,186 @@ TimeSynchronizationServer & TimeSynchronizationServer::Instance()
...
+    // No way to read from the TrustedTimeSource, fall back to default NTP
+    AttemptToGetFallbackNTPTimeFromDelegate();
+}
+
+void TimeSynchronizationServer::OnDone(ReadClient * apReadClient)
+{
+    using namespace chip::app::Clusters::TimeSynchronization::Attributes;
+
+    Granularity::TypeInfo::Type granularity = GranularityEnum::kNoTimeGranularity;
+    mAttributeCache->Get<Granularity::TypeInfo>(kRootEndpointId, granularity);
```
Comment (bzbarsky-apple) [inline src/app/clusters/time-synchronization-server/time-synchronization-server.cpp:351]: This is assuming that on failure the out param will not be modified, no?  Nothing in the API guarantees that.

File: src/app/clusters/time-synchronization-server/time-synchronization-server.h [hunk 5]
```diff
@@ -96,20 +110,53 @@ class TimeSynchronizationServer : public FabricTable::Delegate
...
+    GranularityEnum mGranularity = GranularityEnum::kNoTimeGranularity;
 
     TimeSyncDataProvider::TimeZoneStore mTz[CHIP_CONFIG_TIME_ZONE_LIST_MAX_SIZE];
     Structs::DSTOffsetStruct::Type mDst[CHIP_CONFIG_DST_OFFSET_LIST_MAX_SIZE];
 
     TimeSyncDataProvider mTimeSyncDataProvider;
     static TimeSynchronizationServer sTimeSyncInstance;
     TimeSyncEventFlag mEventFlag = TimeSyncEventFlag::kNone;
+#if TIME_SYNC_ENABLE_TSC_FEATURE
+    Platform::UniquePtr<app::ClusterStateCache> mAttributeCache;
```
Comment (bzbarsky-apple) [inline src/app/clusters/time-synchronization-server/time-synchronization-server.h:145]: So once we read the values... why do we hold on to them forever?  Again, this is not great for constrained devices...

File: src/app/clusters/time-synchronization-server/time-synchronization-server.cpp (hunk 1 above)
Comment (cecille) [reply src/app/clusters/time-synchronization-server/time-synchronization-server.cpp:288]: It shouldn't but I figured it wouldn't hurt to check. Checking span sizes generally just seems like good practice.

File: src/app/chip_data_model.gni [hunk 6]
```diff
@@ -257,6 +262,8 @@ template("chip_data_model") {
           "${_app_root}/clusters/${cluster}/DefaultTimeSyncDelegate.cpp",
           "${_app_root}/clusters/${cluster}/TimeSyncDataProvider.cpp",
         ]
+        cflags +=
+            [ "-DTIME_SYNC_ENABLE_TSC_FEATURE=${time_sync_enable_tsc_feature}" ]
```
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/app/chip_data_model.gni:266]: Addressed in https://github.com/project-chip/connectedhomeip/pull/29396

File: src/app/clusters/time-synchronization-server/DefaultTimeSyncDelegate.h [hunk 7]
```diff
@@ -33,6 +33,9 @@ class DefaultTimeSyncDelegate : public Delegate
     bool HandleUpdateDSTOffset(CharSpan name) override;
     bool IsNTPAddressValid(CharSpan ntp) override;
     bool IsNTPAddressDomain(CharSpan ntp) override;
+    CHIP_ERROR UpdateTimeFromPlatformSource(chip::Callback::Callback<OnTimeSyncCompletion> * callback) override;
+    CHIP_ERROR UpdateTimeUsingNTPFallback(const CharSpan & fallbackNTP,
```
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/app/clusters/time-synchronization-server/DefaultTimeSyncDelegate.h:37]: Addressed in https://github.com/project-chip/connectedhomeip/pull/29396

File: src/app/clusters/time-synchronization-server/time-synchronization-delegate.h [hunk 8]
```diff
@@ -74,6 +77,30 @@ class Delegate
      */
     virtual bool IsNTPAddressDomain(const CharSpan ntp) = 0;
 
+    /**
+     * @brief Delegate should attempt to get time from a platform-defined source using the ordering defined in the
+     * Time source prioritization spec section. Delegate may skip any unsupported sources
+     * Order: GNSS -> trusted high-resolution external source (PTP, trusted network NTP, cloud) ->
+     * local network defined NTP (DHCPv6 -> DHCP -> DNS-SD sources)
+     * If the delegate is unable to support any source, it may return an error immediately. If the delegate is going
+     * to attempt to obtain time from any source, it returns CHIP_NO_ERROR and calls the callback on completion.
```
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/app/clusters/time-synchronization-server/time-synchronization-delegate.h:86]: Addressed in https://github.com/project-chip/connectedhomeip/pull/29396

File: src/app/clusters/time-synchronization-server/time-synchronization-delegate.h [hunk 9]
```diff
@@ -74,6 +77,30 @@ class Delegate
...
+     * If the delegate successfully obtains the time, it sets the time using the platform time API (SetClock_RealTime)
+     * and calls the callback with the time source and granularity set as appropriate.
+     * If the delegate is unsuccessful in obtaining the time, it calls the callback with timeSource set to kNone and
+     * granularity set to kNoTimeGranularity.
+     */
+    virtual CHIP_ERROR UpdateTimeFromPlatformSource(chip::Callback::Callback<OnTimeSyncCompletion> * callback) = 0;
```
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/app/clusters/time-synchronization-server/time-synchronization-delegate.h:92]: Addressed in https://github.com/project-chip/connectedhomeip/pull/29396

File: src/app/clusters/time-synchronization-server/time-synchronization-server.cpp (hunk 1 above)
Comment (bzbarsky-apple) [reply src/app/clusters/time-synchronization-server/time-synchronization-server.cpp:288]: I guess it makes it look like if the size check fails you are "ok" whereas in fact you've already been owned... but ok.

File: src/app/clusters/time-synchronization-server/time-synchronization-server.cpp (hunk 2 above)
Comment (bzbarsky-apple) [reply src/app/clusters/time-synchronization-server/time-synchronization-server.cpp:303]: Addressed in https://github.com/project-chip/connectedhomeip/pull/29396

File: src/app/clusters/time-synchronization-server/time-synchronization-server.cpp (hunk 3 above)
Comment (bzbarsky-apple) [reply src/app/clusters/time-synchronization-server/time-synchronization-server.cpp:314]: Addressed in https://github.com/project-chip/connectedhomeip/pull/29396

File: src/app/clusters/time-synchronization-server/time-synchronization-server.cpp (hunk 4 above)
Comment (bzbarsky-apple) [reply src/app/clusters/time-synchronization-server/time-synchronization-server.cpp:351]: Addressed in https://github.com/project-chip/connectedhomeip/pull/29396

File: src/app/clusters/time-synchronization-server/time-synchronization-server.h (hunk 5 above)
Comment (bzbarsky-apple) [reply src/app/clusters/time-synchronization-server/time-synchronization-server.h:145]: Addressed in https://github.com/project-chip/connectedhomeip/pull/29396

**Pull request merged**